File: frontend/web/src/services/unlockRecord.service.js
import { httpClient } from './http.client';
import { API_ENDPOINTS } from './api.config';

/**
 * 解锁记录服务
 */
class UnlockRecordService {
  /**
   * 创建解锁记录（支付成功后调用）
   * @param {Object} data - 解锁记录数据
   * @param {number} data.resume_id - 简历 ID
   * @param {number} data.buyer_id - 购买者用户 ID
   * @param {string} data.buyer_wallet - 购买者钱包地址
   * @param {string} data.seller_wallet - 卖家钱包地址
   * @param {number} data.amount - 支付金额
   * @param {string} data.transaction_signature - 交易签名（Sui transaction digest）
   * @param {number} [data.block_time] - 区块时间戳（可选）
   * @returns {Promise<Object>} 创建的解锁记录
   */
  async createUnlockRecord(data) {
    try {
      const response = await httpClient.post(API_ENDPOINTS.unlockRecords.create, data);
      return response.data;
    } catch (error) {
      console.error('创建解锁记录失败:', error);
      throw error;
    }
  }

  /**
   * 检查用户是否已解锁某简历
   * @param {number} resumeId - 简历 ID
   * @param {number} buyerId - 购买者用户 ID
   * @returns {Promise<boolean>} 是否已解锁
   */
  async checkUnlockStatus(resumeId, buyerId) {
    try {
      const response = await httpClient.get(
        API_ENDPOINTS.unlockRecords.checkUnlock(resumeId, buyerId)
      );
      return response.data?.unlocked || false;
    } catch (error) {
      console.error('检查解锁状态失败:', error);
      return false;
    }
  }

  /**
   * 获取用户已解锁的所有简历
   * @param {string} buyerWallet - 购买者钱包地址
   * @returns {Promise<Array>} 解锁记录列表
   */
  async getUnlockedResumes(buyerWallet) {
    try {
      const response = await httpClient.get(
        API_ENDPOINTS.unlockRecords.getUnlockedByBuyer(buyerWallet)
      );
      return response.data || [];
    } catch (error) {
      console.error('获取已解锁简历失败:', error);
      return [];
    }
  }

  /**
   * 获取简历的所有解锁记录（简历所有者查看）
   * @param {number} resumeId - 简历 ID
   * @returns {Promise<Array>} 解锁记录列表
   */
  async getResumeUnlockRecords(resumeId) {
    try {
      const response = await httpClient.get(
        API_ENDPOINTS.unlockRecords.getByResume(resumeId)
      );
      return response.data || [];
    } catch (error) {
      console.error('获取简历解锁记录失败:', error);
      return [];
    }
  }
}

export default new UnlockRecordService();

File: backend/rust_backend/src/services/unlock_record_service.rs
use sea_orm::DatabaseConnection;
use crate::dao::UnlockRecordDao;
use crate::entities::unlock_record;

pub struct UnlockRecordService;

impl UnlockRecordService {
    /// 创建解锁记录（支付后调用）
    pub async fn create_unlock_record(
        db: &DatabaseConnection,
        resume_id: i64,
        buyer_id: i64,
        buyer_wallet: String,
        seller_wallet: String,
        amount: i64,
        transaction_signature: String,
        block_time: Option<i64>,
    ) -> Result<unlock_record::Model, String> {
        // 检查是否已存在相同的交易签名（防止重复记录）
        let exists = UnlockRecordDao::exists_by_signature(db, &transaction_signature)
            .await
            .map_err(|e| format!("Failed to check transaction signature: {}", e))?;

        if exists {
            return Err("Transaction already recorded".to_string());
        }

        // 创建解锁记录
        UnlockRecordDao::create(
            db,
            resume_id,
            buyer_id,
            buyer_wallet,
            seller_wallet,
            amount,
            transaction_signature,
            block_time,
        )
        .await
        .map_err(|e| format!("Failed to create unlock record: {}", e))
    }

    /// 检查用户是否已解锁某简历
    pub async fn has_unlocked(
        db: &DatabaseConnection,
        resume_id: i64,
        buyer_id: i64,
    ) -> Result<bool, String> {
        let record = UnlockRecordDao::find_by_resume_and_buyer(db, resume_id, buyer_id)
            .await
            .map_err(|e| format!("Failed to check unlock status: {}", e))?;

        Ok(record.is_some())
    }

    /// 获取用户已解锁的所有简历
    pub async fn get_unlocked_resumes(
        db: &DatabaseConnection,
        buyer_wallet: &str,
    ) -> Result<Vec<unlock_record::Model>, String> {
        UnlockRecordDao::find_unlocked_resumes_by_buyer(db, buyer_wallet)
            .await
            .map_err(|e| format!("Failed to get unlocked resumes: {}", e))
    }

    /// 获取简历的所有解锁记录（简历所有者可查看）
    pub async fn get_resume_unlock_records(
        db: &DatabaseConnection,
        resume_id: i64,
    ) -> Result<Vec<unlock_record::Model>, String> {
        UnlockRecordDao::find_by_resume_id(db, resume_id)
            .await
            .map_err(|e| format!("Failed to get unlock records: {}", e))
    }
}
